[fetch] Implement the `Response.json` static method

andreubotella opened the pull request:
#### 206817f17f39b6de35ea63268b74b750030d574f
<pre>
[fetch] Implement the `Response.json` static method
<a href="https://bugs.webkit.org/show_bug.cgi?id=240375">https://bugs.webkit.org/show_bug.cgi?id=240375</a>

Reviewed by Youenn Fablet.

This implements the `Response.json` static method, added to the fetch
spec in <a href="https://github.com/whatwg/fetch/pull/1392.">https://github.com/whatwg/fetch/pull/1392.</a>

* LayoutTests/imported/w3c/web-platform-tests/fetch/api/response/response-static-json.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/response/response-static-json.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/response/response-static-json.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/response/response-static-json.any.worker-expected.txt:
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::create):
  Added this method overload to implement the spec&apos;s &quot;initialize a
  response&quot; algorithm. This algortihm used to be combined with the
  regular Response creation algorithm which extracts the body, but
  Response.json cannot use that directly.
(WebCore::FetchResponse::staticJson):
* Source/WebCore/Modules/fetch/FetchResponse.h:
* Source/WebCore/Modules/fetch/FetchResponse.idl:

Canonical link: <a href="https://commits.webkit.org/261960@main">https://commits.webkit.org/261960@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5fcd50d6371618768f60a163e5d9f70e429c7782

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113237 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22370 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1910 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/5012 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121691 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117370 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23740 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13543 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/6237 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/119024 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17636 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100900 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106332 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99589 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/1441 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46682 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14669 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1480 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/98943 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10804 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20671 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53471 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17231 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/fetch/FetchResponse.cpp:
```diff
@@ -168,6 +163,21 @@ ExceptionOr<Ref<FetchResponse>> FetchResponse::redirect(ScriptExecutionContext&
+        return Exception { TypeError, "Value doesn't have a JSON representation"_s };
+
+    String ignored;
+    auto bodyResult = FetchBody::extract(WTFMove(jsonString), ignored);
```
This is a no-op in our implementation, I would think we do not need this since body can be a String.

andreubotella replied:
`FetchBody` doesn't have any public constructors, and it seems like the only ways to create an instance are the `extract` and `fromFormData` static methods.

youennf replied:
Let's move the String constructor to public.

youennf commented on Source/WebCore/Modules/fetch/FetchResponse.cpp:
```diff
@@ -168,6 +163,21 @@ ExceptionOr<Ref<FetchResponse>> FetchResponse::redirect(ScriptExecutionContext&
+
+    String ignored;
+    auto bodyResult = FetchBody::extract(WTFMove(jsonString), ignored);
+    if (bodyResult.hasException())
```
In this code path, there should be no exception AIUI, can you check this?

youennf commented on Source/WebCore/Modules/fetch/FetchResponse.idl:
```diff
@@ -48,6 +48,7 @@ dictionary FetchResponseInit {
 
     [CallWith=CurrentScriptExecutionContext, NewObject] static FetchResponse error();
     [CallWith=CurrentScriptExecutionContext, NewObject] static FetchResponse redirect(USVString url, optional unsigned short status = 302);
+    [CallWith=CurrentScriptExecutionContext, NewObject, ImplementedAs=staticJson] static FetchResponse json(any data, optional FetchResponseInit init);
```
Maybe jsonForBindings?

youennf reviewed (approved):
This looks ok to me.
A few potential improvements below.

andreubotella committed:
[fetch] Implement the `Response.json` static method
https://bugs.webkit.org/show_bug.cgi?id=240375

Reviewed by Youenn Fablet.

This implements the `Response.json` static method, added to the fetch
spec in https://github.com/whatwg/fetch/pull/1392.

* LayoutTests/imported/w3c/web-platform-tests/fetch/api/response/response-static-json.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/response/response-static-json.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/response/response-static-json.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fetch/api/response/response-static-json.any.worker-expected.txt:
* Source/WebCore/Modules/fetch/FetchBody.h:
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::create):
  Added this method overload to implement the spec's "initialize a
  response" algorithm. This algortihm used to be combined with the
  regular Response creation algorithm which extracts the body, but
  Response.json cannot use that directly.
(WebCore::FetchResponse::staticJson):
* Source/WebCore/Modules/fetch/FetchResponse.h:
* Source/WebCore/Modules/fetch/FetchResponse.idl:

Canonical link: https://commits.webkit.org/261960@main
